mfr/extensions/tabular/libs/stdlib_tools.py

```diff
@@ -92,3 +110,63 @@ def _set_dialect_quote_attrs(dialect, data):
             dialect.quotechar = '"'
         if re.search('"""[[({]\'.+\',', data):
             dialect.doublequote = True
+
+
+def _trim_or_append_data(fp, text, read_size, size_to_sniff, max_render_size=MAX_FILE_SIZE):
+    """Recursively read data from a file and return its full first row.  The file starts with
+    ``text`` and the file pointer points to the next character immediately after `text`.
+
+    :param fp: the file pointer from which data is read
+    :param text: the current text chunk to check the new line character
+    :param read_size: the last read size when `fp.read()` is called
+    :param size_to_sniff: the accumulated size fo the text to sniff
+    :param max_render_size: the max file size for render
+    :return: the first row of the file in string
+    """
+
+    # Return on empty text. This handles the corner case where the CSV is empty or only contains
+    # one line without any new line characters.
+    if len(text) == 0:
+        return ''
+
+    # Try to find the first new line character in the text chunk
+    index = _find_new_line(text)
+    # If found, return the trimmed substring
+    if index != -1:
+        return text[:index]
+    # Otherwise, update `sniff_size` and then sniff more (2 times of the last `read_size`) text
+    size_to_sniff += read_size
+    read_size *= 2
+    more_text = fp.read(read_size)
+
+    # If text to sniff now goes over the max file size limit, raise the renderer error since there
+    # is no need to sniff when the file is already too large to be rendered.
+    if size_to_sniff + len(more_text) >= max_render_size:
+        raise TabularRendererError(
+            'The first row of this file is too large for the sniffer to detect the dialect. '
+            'Please download and view it locally.',
+            code=400,
+            extension='csv'
+        )
+    # If the size is still within the limit, recursively check `more_text`
+    return text + _trim_or_append_data(fp, more_text, read_size, size_to_sniff,
+                                       max_render_size=max_render_size)
+
+
+def _find_new_line(text):
+    """In the given text string, find the index of the first occurrence of any of the three types
+    of new line character. Note: '\n\r' is not a new line character but two, one LF and one CR.
+
+    1. \r\n     Carriage Return (CR) and Line Feed (LF), must be checked first
+    2. \n       LF
+    3. \r       CR
+
+    :param text: the text string to check
+    :return: the index of the first new line character if found. Otherwise, return -1.
+    """
+    index = text.find('\r\n')
+    if index == -1:
+        index = text.find('\n')
+        if index == -1:
+            index = text.find('\r')
```

Comment:
I am not sure why `.rfind()` was used in the initial solution. One guess is that it only reads a fixed size for sniff so it wants to include as many role as possible. However, this is no longer the case in my implementation where only the first row is sniffed.

```python
    index = text.rfind('\r\n')
    if index == -1:
        index = text.rfind('\n')
        if index == -1:
            index = text.rfind('\r')
```